[feat] Add SQLite support (better-sqlite3)

samuelmbabhazi opened the pull request:
### Before submitting the PR, please make sure you do the following

1.  Contributor license agreement
    For us it's important to have the agreement of our contributors to use their work, whether it be code or documentation. Therefore, we are asking all contributors to sign a contributor license agreement (CLA) as commonly accepted in most open source projects. **Just open the pull request and our CLA bot will prompt you briefly.**

2.  Please check our [contribution guidelines](https://docs.traduora.co/docs/contributing#review-process-for-this-repo) for some help in the process.

samuelmbabhazi committed:
feat: Add SQLite support

samuelmbabhazi committed:
feat: remove sqlite file

samuelmbabhazi committed:
fix datasource config

samuelmbabhazi committed:
Remove unnecessary changes

samuelmbabhazi committed:
Refactor datasource

samuelmbabhazi committed:
Merge branch 'develop' into feat/468-add-sqlite-support

samuelmbabhazi committed:
feat(Sqlite):update migrations

samuelmbabhazi committed:
feat(Sqlite):fix import

samuelmbabhazi committed:
feat(Sqlite):fix deepScan

samuelmbabhazi committed:
fix migration

samuelmbabhazi committed:
fix import

samuelmbabhazi committed:
remove uncessary changes

samuelmbabhazi committed:
fix import

samuelmbabhazi committed:
fix import

samuelmbabhazi committed:
feat(Sqlite):fix deepScan

samuelmbabhazi committed:
fix migrations

samuelmbabhazi committed:
Remap exception

samuelmbabhazi committed:
Make default Mysql

samuelmbabhazi committed:
feedback integration

samuelmbabhazi committed:
refactor db util

samuelmbabhazi committed:
AI Suggestion Integrations

samuelmbabhazi commented:
@evereq, here are some videos of the tests done with the 3 DBMSs to ensure that nothing was broken and everything is working properly.

https://www.loom.com/share/38c9f707de964365bd09d3ca536a4e62?sid=4a3c8838-b509-442f-9947-f93433b101fe

https://www.loom.com/share/6879cc827f484e8cb47cc5f163c667eb?sid=5402f857-13b3-454c-9903-e3273c23d309

https://www.loom.com/share/b8d5e813b592411dab2e90c11c791cde?sid=12fbcd86-325f-43c9-af07-fa1ec96d6bff

samuelmbabhazi commented on api/src/migrations/1537801450876-add-project-user-role.ts:
```diff
@@ -1,29 +1,58 @@
+        await queryRunner.query('ALTER TABLE "project_user" DROP COLUMN "role"');
+        break;
       default:
         console.log('Unknown DB type');
```
this is not necessary at the moment

samuelmbabhazi committed:
AI Suggestion Integrations

samuelmbabhazi commented on api/src/utils/database-type-helper.ts:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * Type for encrypted secret columns
+   */
+  encryptedSecret: (): ColumnOptions => {
+    return getDbColumnType({ postgres: 'bytea', mysql: 'binary', betterSqlite3: 'blob' }, { ...(isDbType(DbType.MYSQL) ? { length: 60 } : {}) });
+  },
```
certain property will be passed when called depending on the context

samuelmbabhazi committed:
AI Suggestion Integrations

samuelmbabhazi committed:
AI Suggestion Integrations

dougg0k commented:
I would recommend adding another docker-compose for the sqlite version, where it does not include another db, only the app.  Where the app should auto create the sqlite db. 

Your decision whether you want the db filename and path to be set via envvar or not.

samuelmbabhazi committed:
feat: add SQLite support as default database option in Docker Compose

- Renamed original docker-compose.yaml to docker-compose.mysql.yaml
- Created new docker-compose.yaml with better-sqlite3 as default database
- Updated docker-compose.demo.yaml to use better-sqlite3 instead of MySQL
- Added volume mapping for SQLite database persistence at ./traduora_sqlite_data:/app/api/data
- Added support for environment variable TR_DB_SQLITE_PATH with default path
- Organized Docker Compose files to support multiple database options:
  * SQLite: docker-compose.yaml (default) & docker-compose.demo.yaml
  * MySQL: docker-compose.mysql.yaml
  * PostgreSQL: docker-compose.postgres.yaml

evereq commented:
@samuelmbabhazi did you runned all tests locally? I.e. e2e / unit etc, all works?

dougg0k commented:
You guys do these things locally? Should be very easy to automate / setup tests among other things with github actions.

Edit:

nvm, you have circleci, but maybe not for PRs.

samuelmbabhazi commented:
> @samuelmbabhazi did you runned all tests locally? I.e. e2e / unit etc, all works?

yes sir @evereq the tests are working

samuelmbabhazi commented on api/src/migrations/1537531930470-add locales.ts:
```diff
@@ -1149,6 +1150,585 @@ export class addLocales1537531930470 implements MigrationInterface {
+                ('bo_IN', 'Tibetan', 'India'),
+                ('ti', 'Tigrinya', ''),
+                ('ti_ER', 'Tigrinya', 'Eritrea'),
+                ('ti_ET', 'Tigrinya', 'Ethiopia'),
+                ('to', 'Tongan', ''),
+                ('to_TO', 'Tongan', 'Tonga'),
+                ('tr', 'Turkish', ''),
+                ('tr_CY', 'Turkish', 'Cyprus'),
+                ('tr_TR', 'Turkish', 'Turkey'),
+                ('uk', 'Ukrainian', ''),
+                ('uk_UA', 'Ukrainian', 'Ukraine'),
+                ('ur', 'Urdu', ''),
+                ('ur_IN', 'Urdu', 'India'),
+                ('ur_PK', 'Urdu', 'Pakistan'),
+                ('ug', 'Uyghur', ''),
+                ('ug_Arab_CN', 'Uyghur', 'Arabic, China'),
+                ('ug_Arab', 'Uyghur', 'Arabic'),
+                ('ug_CN', 'Uyghur', 'China'),
+                ('uz', 'Uzbek', ''),
+                ('uz_AF', 'Uzbek', 'Afghanistan'),
+                ('uz_Arab_AF', 'Uzbek', 'Arabic, Afghanistan'),
+                ('uz_Arab', 'Uzbek', 'Arabic'),
+                ('uz_Cyrl_UZ', 'Uzbek', 'Cyrillic, Uzbekistan'),
+                ('uz_Cyrl', 'Uzbek', 'Cyrillic'),
+                ('uz_Latn_UZ', 'Uzbek', 'Latin, Uzbekistan'),
+                ('uz_Latn', 'Uzbek', 'Latin'),
+                ('uz_UZ', 'Uzbek', 'Uzbekistan'),
+                ('vi', 'Vietnamese', ''),
+                ('vi_VN', 'Vietnamese', 'Vietnam'),
+                ('cy', 'Welsh', ''),
+                ('cy_GB', 'Welsh', 'United Kingdom'),
+                ('fy', 'Western Frisian', ''),
+                ('fy_NL', 'Western Frisian', 'Netherlands'),
+                ('yi', 'Yiddish', ''),
+                ('yo', 'Yoruba', ''),
+                ('yo_BJ', 'Yoruba', 'Benin'),
+                ('yo_NG', 'Yoruba', 'Nigeria'),
+                ('zu', 'Zulu', ''),
+                ('zu_ZA', 'Zulu', 'South Africa')
+        `);
```
I will do this in another specific PR since it is a major change.

samuelmbabhazi commented on api/src/migrations/1544283791233-add-tos-and-privacy-fields.ts:
```diff
@@ -1,32 +1,84 @@
     switch (config.db.default.type) {
-      case 'postgres':
+      case DbType.POSTGRES:
         await queryRunner.query(`ALTER TABLE "user" ADD "tos_and_privacy_accepted_date" timestamp NULL;`);
         await queryRunner.query(`ALTER TABLE "user" ADD COLUMN "tos_and_privacy_accepted_version" varchar(255) NULL;`);
         break;
-      case 'mysql':
+      case DbType.MYSQL:
         await queryRunner.query('ALTER TABLE `user` ADD `tosAndPrivacyAcceptedDate` timestamp(6) NULL');
         await queryRunner.query('ALTER TABLE `user` ADD `tosAndPrivacyAcceptedVersion` varchar(255) NULL');
         break;
+      case DbType.BETTER_SQLITE3:
+        await queryRunner.query('ALTER TABLE "user" ADD COLUMN "tos_and_privacy_accepted_date" DATETIME NULL');
+        await queryRunner.query('ALTER TABLE "user" ADD COLUMN "tos_and_privacy_accepted_version" TEXT NULL');
```
For the moment this is not a problem, so let's keep it as such

samuelmbabhazi commented on api/src/migrations/1562578811334-provider google.ts:
```diff
@@ -1,29 +1,42 @@
+        await queryRunner.query('ALTER TABLE "user" RENAME COLUMN "encrypted_password_temp" TO "encrypted_password"');
+        break;
       default:
         console.log('Unknown DB type');
```
no need to shut down the application for this

samuelmbabhazi commented on api/src/migrations/1667573768424-add-term-context.ts:
```diff
@@ -1,29 +1,35 @@
+      case DbType.BETTER_SQLITE3:
+        await queryRunner.query('ALTER TABLE "term" ADD COLUMN "context" TEXT DEFAULT NULL');
+        break;
       default:
         console.log('Unknown DB type');
     }
```
no need to shut down the application for this

samuelmbabhazi commented on api/src/migrations/1537801450876-add-project-user-role.ts:
```diff
@@ -1,29 +1,54 @@
+        await queryRunner.query(`DROP TABLE "project_user_temp"`);
+        break;
       default:
         console.log('Unknown DB type');
```
we don't need to shut down the app for that

samuelmbabhazi commented on api/src/migrations/1542044660604-fix-translations-primary-key.ts:
```diff
@@ -1,39 +1,75 @@
+        break;
       default:
         console.log('Unknown DB type');
     }
```
we don't need to shut down the app for that

samuelmbabhazi commented on api/src/migrations/1549613347230-project-users-index.ts:
```diff
@@ -1,28 +1,35 @@
+      case DbType.BETTER_SQLITE3:
+        await queryRunner.query('CREATE UNIQUE INDEX "IDX_20543d6caa7324ce6706fad2f5" ON "project_user"("project_id", "user_id")');
+        break;
       default:
         console.log('Unknown DB type');
     }
```
we don't need to shut down the app for that

evereq replied:
@samuelmbabhazi we actually do need, because if wrong/unsupported DB type was set, APP should fail with error!

samuelmbabhazi commented on api/src/migrations/1551022480406-per-user-project-limits.ts:
```diff
@@ -1,15 +1,19 @@
         break;
-      case 'mysql':
+      case DbType.MYSQL:
         await queryRunner.query('ALTER TABLE `user` ADD `numProjectsCreated` int NOT NULL DEFAULT 0');
```
everything is managed for this difference in nomenclature

samuelmbabhazi commented on api/src/migrations/1562578811334-provider google.ts:
```diff
@@ -1,29 +1,42 @@
         break;
-      case 'mysql':
+      case DbType.MYSQL:
         await queryRunner.query('ALTER TABLE `user` CHANGE `encryptedPassword` `encryptedPassword` binary(60) NULL');
```
everything is managed for this difference in nomenclature

samuelmbabhazi commented on api/src/migrations/1542044660604-fix-translations-primary-key.ts:
```diff
@@ -1,39 +1,78 @@
-        await queryRunner.query(`ALTER TABLE "translation" DROP CONSTRAINT "translation_pkey";
-    `);
-        await queryRunner.query(`ALTER TABLE "translation" DROP COLUMN "id";`);
-        await queryRunner.query(`ALTER TABLE "translation" ADD PRIMARY KEY ("term_id", "project_locale_id");`);
+      case DbType.POSTGRES:
```
This is not true

evereq committed:
Update api/src/connection/datasource.ts

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

evereq committed:
Update api/src/migrations/1537801450876-add-project-user-role.ts

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

evereq committed:
Update api/src/migrations/1542044660604-fix-translations-primary-key.ts

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

evereq committed:
Update api/src/migrations/1543494409127-change-translation-value-type.ts

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

evereq commented on api/src/migrations/1544283791233-add-tos-and-privacy-fields.ts:
```diff
@@ -1,32 +1,84 @@
+          
+          BEGIN TRANSACTION;
+          
+          -- Créer une table temporaire sans les colonnes à supprimer
+          CREATE TABLE user_temp (
```
@samuelmbabhazi  why comment in French!? Please make sure you keep all in English. Also how you missed that suggestion? Let's fix in next PR please!

evereq committed:
Update api/src/migrations/1549613347230-project-users-index.ts

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

evereq commented on api/src/migrations/1549613347230-project-users-index.ts:
```diff
@@ -1,28 +1,35 @@
         await queryRunner.query('DROP INDEX `IDX_20543d6caa7324ce6706fad2f5` ON `project_user`');
         break;
+      case DbType.BETTER_SQLITE3:
+        await queryRunner.query('DROP INDEX IF EXISTS "IDX_20543d6caa7324ce6706fad2f5"');
```
@samuelmbabhazi did you verified this? Is it required!?

evereq commented on api/src/migrations/1552494719664-remove-tos-and-privacy.ts:
```diff
@@ -1,32 +1,41 @@
         await queryRunner.query('ALTER TABLE `user` ADD `tosAndPrivacyAcceptedDate` timestamp(6) NULL');
         break;
+      case DbType.BETTER_SQLITE3:
+        await queryRunner.query('ALTER TABLE "user" ADD COLUMN "tos_and_privacy_accepted_version" TEXT NULL');
+        await queryRunner.query('ALTER TABLE "user" ADD COLUMN "tos_and_privacy_accepted_date" TEXT NULL');
```
`tos_and_privacy_accepted_date` type TEXT!?? @samuelmbabhazi How you mising bot suggestions!? It's for sure best to use DATETIME for dates, not TEXT

evereq commented on api/src/migrations/1557936309231-add-invite.ts:
```diff
@@ -1,63 +1,80 @@
+            "email" TEXT NOT NULL,
+            "status" TEXT NOT NULL DEFAULT 'sent' CHECK ("status" IN ('sent', 'accepted')),
+            "role" TEXT NOT NULL DEFAULT 'viewer' CHECK ("role" IN ('admin', 'editor', 'viewer')),
+            "project_id" TEXT NOT NULL,
```
@samuelmbabhazi ^^^

evereq commented:
@samuelmbabhazi you missed a LOT of bot suggestions. I merged some, but still tons of missed. Please go over all and create new PR based on develop branch with more fixes if needed.
